lib/util/DateTimeHelpers.ts

```diff
@@ -0,0 +1,191 @@
+  return `${numSer(time.hours)}:${numSer(time.minutes)}:${numSer(time.seconds)}${timeZoneSerializer(time)}`;
+}
+
+export function durationParser(durationStr: string): Partial<IDurationRepresentation> {
+  // https://www.w3.org/TR/xmlschema-2/#duration-lexical-repr
+  const [ dayNotation, timeNotation ] = durationStr.split('T');
+
+  // Handle date part
+  const formattedDayDur = dayNotation.replace(/^(-)?P(\d+Y)?(\d+M)?(\d+D)?$/gu, '$11S!$2!$3!$4');
+  if (formattedDayDur === dayNotation) {
+    throw new Error('nono');
+  }
+
+  const durationStrings = formattedDayDur.split('!');
+  if (timeNotation !== undefined) {
+    const formattedTimeDur = timeNotation.replace(/^(\d+H)?(\d+M)?(\d+(\.\d+)?S)?$/gu, '$1!$2!$3');
+
+    if (timeNotation === '' || timeNotation === formattedTimeDur) {
+      throw new Error('nono');
+    }
+    durationStrings.push(...formattedTimeDur.split('!'));
+  }
+  const duration = durationStrings.map(str => str.slice(0, -1));
+  if (!duration.slice(1).some(item => item)) {
+    throw new Error('nono');
+  }
+
+  const sign = <-1 | 1> Number(duration[0]);
+  return {
+    year: duration[1] ? sign * Number(duration[1]) : undefined,
+    month: duration[2] ? sign * Number(duration[2]) : undefined,
+    day: duration[3] ? sign * Number(duration[3]) : undefined,
+    hours: duration[4] ? sign * Number(duration[4]) : undefined,
+    minutes: duration[5] ? sign * Number(duration[5]) : undefined,
+    seconds: duration[6] ? sign * Number(duration[6]) : undefined,
+  };
+}
+
+export function yearMonthDurationParser(durationStr: string): Partial<IYearMonthDurationRepresentation> {
+  const res = durationParser(durationStr);
+  if (Object.entries(res).some(([ key, value ]) => value !== undefined && ![ 'year', 'month' ].includes(key))) {
+    throw new Error('nono');
+  }
+  return res;
+}
+
+export function dayTimeDurationParser(durationStr: string): Partial<IDayTimeDurationRepresentation> {
+  const res = durationParser(durationStr);
+  const filtered = extractDayTimeDur(res);
+  if (Object.entries(res).some(([ key, value ]) => value !== undefined && ![
+    'hours',
+    'minutes',
+    'seconds',
+    'day',
+  ].includes(key))) {
+    throw new Error('nono');
+  }
+  return res;
+}
+
+export function durationSerializer(dur: Partial<IDurationRepresentation>, zeroString: 'PT0S' | 'P0M' = 'PT0S'): string {
+  if (!Object.values(dur).some(val => (val || 0) !== 0)) {
+    return zeroString;
+  }
+  const negative: boolean = Object.values(dur).some(val => (val || 0) < 0);
+  const hasTimeField = !!(dur.hours || dur.minutes || dur.seconds);
+  return `${negative ? '-' : ''}P${dur.year ? `${Math.abs(dur.year)}Y` : ''}${dur.month ?
+    `${Math.abs(dur.month)}M` :
+    ''}${dur.day ?
+    `${Math.abs(dur.day)}D` :
+    ''}${hasTimeField ?
+    `T${dur.hours ? `${Math.abs(dur.hours)}H` : ''}${dur.minutes ?
+      `${Math.abs(dur.minutes)}M` :
+      ''}${dur.seconds ? `${Math.abs(dur.seconds)}S` : ''}` :
+    ''}`;
+}
```

Comment:
This would need some comments and/or more clear code. A lot is going on at once. An example input/output might also be very nice.

Reply:
I don't think I can provide cleaner code. I't just some string formatting I'm afraid. I can however give an example input/ output.

Reply:
Just extracting the parts into separate statements would help.
E.g.
```
const sign = negative ? '-' : '';
const year = dur.year ? `${Math.abs(dur.year)}Y` : ''
# ... etc
```
Especially since the linebreaks are in weird places for readability.